File: src/main/java/com/evalvis/ratelimiter/key/ClientIpResolver.java
package com.evalvis.ratelimiter.key;

import jakarta.servlet.http.HttpServletRequest;

public final class ClientIpResolver {

	private ClientIpResolver() {
	}

	public static String resolve(HttpServletRequest request) {
		String forwarded = request.getHeader("X-Forwarded-For");
		if (forwarded != null && !forwarded.isBlank()) {
			int comma = forwarded.indexOf(',');
			String first = comma >= 0 ? forwarded.substring(0, comma) : forwarded;
			return first.trim();
		}
		return request.getRemoteAddr();
	}

}

File: loadbalancer/src/main/kotlin/com/evalvis/loadbalancer/web/ClientIpResolver.kt
package com.evalvis.loadbalancer.web

import jakarta.servlet.http.HttpServletRequest

object ClientIpResolver {

	fun resolve(request: HttpServletRequest): String {
		val xff = request.getHeader("X-Forwarded-For")
		if (!xff.isNullOrBlank()) {
			return xff.split(",").first().trim()
		}
		return request.remoteAddr
	}
}
